### PR TITLE
Grant revoke role

### PR DESCRIPTION
**Pull request by arhimondr**

### CODE REVIEW

File: presto-hive/src/main/java/com/facebook/presto/hive/security/SqlStandardAccessControl.java [hunk 1]
```diff
@@ -331,15 +350,74 @@ private boolean checkTablePermission(ConnectorTransactionHandle transaction, Ide
...
+                .map(RoleGrant::getRoleName)
+                .collect(toSet());
+    }
+
+    private static Set<HivePrivilegeInfo> listApplicableTablePrivileges(SemiTransactionalHiveMetastore metastore, String databaseName, String tableName, PrestoPrincipal principal)
+    {
+        Set<String> applicableRoles = listApplicableRoles(metastore, principal);
+        List<PrestoPrincipal> principals = new ArrayList<>();
+        principals.add(principal);
+        applicableRoles.stream().map(role -> new PrestoPrincipal(ROLE, role)).forEach(principals::add);
```
Comment (amrutagokhale) [inline presto-hive/src/main/java/com/facebook/presto/hive/security/SqlStandardAccessControl.java:467]: If you look at Hive code, they pre-process role names and convert them into lowercase, before passing them to the metastore API. Even though this is a weird requirement, in my experience, the code didn't work otherwise. So if this code path involves calling hive metastore API, probably you should add `toLowercase` in here and all cases where the principal type is role.

File: presto-hive/src/main/java/com/facebook/presto/hive/metastore/ThriftHiveMetastore.java [hunk 2]
```diff
@@ -816,9 +763,9 @@ public void grantTablePrivileges(String databaseName, String tableName, String g
                     .run("grantTablePrivileges", stats.getGrantTablePrivileges().wrap(() -> {
                         try (HiveMetastoreClient metastoreClient = clientProvider.createMetastoreClient()) {
                             HivePrincipal hivePrincipal = toHivePrincipal(grantee);
-                            Set<HivePrivilegeInfo> existingPrivileges = getTablePrivileges(hivePrincipal, databaseName, tableName);
+                            Set<HivePrivilegeInfo> existingPrivileges = listTablePrivileges(databaseName, tableName, new PrestoPrincipal(USER, grantee));
```
Comment (amrutagokhale) [inline presto-hive/src/main/java/com/facebook/presto/hive/metastore/ThriftHiveMetastore.java]: How is `public` handled, if you are passing `USER` as the principal type?

File: presto-hive/src/main/java/com/facebook/presto/hive/metastore/HivePrincipal.java [hunk 3]
```diff
@@ -1,97 +0,0 @@
...
-
-    private final String principalName;
-    private final PrincipalType principalType;
-
-    public HivePrincipal(String principalName, PrincipalType principalType)
-    {
-        this.principalType = principalType;
-
-        if (principalType == ROLE) {
-            this.principalName = principalName.toLowerCase(Locale.US); // Hive metastore API requires role names to be in lowercase.
```
Comment (amrutagokhale) [inline presto-hive/src/main/java/com/facebook/presto/hive/metastore/HivePrincipal.java:41]: If we are deleting this file, there should be a way to ensure that hive role names are always converted to lowercase, before invoking Hive metastore API methods. Or is it the case that you have taken care of this in another way?

File: presto-product-tests/src/main/java/com/facebook/presto/tests/hive/TestRoles.java [hunk 4]
```diff
@@ -137,8 +138,198 @@ public void testAccessControl()
...
+        QueryAssert.assertThat(onPrestoBob().executeQuery("SELECT * FROM hive.information_schema.applicable_roles"))
+                .contains(row("bob", "USER", "public", "NO"));
+    }
+
+    @Test(groups = {HIVE_CONNECTOR, ROLES, AUTHORIZATION, PROFILE_SPECIFIC_TESTS})
+    public void testAdminRoleIsGrantedToHive()
+            throws Exception
+    {
+        QueryAssert.assertThat(onPresto().executeQuery("SELECT * FROM hive.information_schema.applicable_roles"))
+                .contains(row("hdfs", "USER", "admin", "YES"));
```
Comment (amrutagokhale) [inline presto-product-tests/src/main/java/com/facebook/presto/tests/hive/TestRoles.java]: either change "hdfs" ->  "hive" or change the function name to `testAdminRoleIsGrantedToHdfs`

File: presto-main/src/main/java/com/facebook/presto/connector/informationSchema/InformationSchemaMetadata.java [hunk 5]
```diff
@@ -99,6 +100,12 @@
             .table(tableMetadataBuilder(TABLE_ROLES)
                     .column("role_name", createUnboundedVarcharType())
                     .build())
+            .table(tableMetadataBuilder(TABLE_APPLICABLE_ROLES)
+                    .column("grantee", createUnboundedVarcharType())
+                    .column("grantee_type", createUnboundedVarcharType())
+                    .column("role_name", createUnboundedVarcharType())
+                    .column("is_grantable", createUnboundedVarcharType())
```
Comment (amrutagokhale) [inline presto-main/src/main/java/com/facebook/presto/connector/informationSchema/InformationSchemaMetadata.java:108]: Make this column to be of `Boolean` type, rather than `varchar`.

**Review by amrutagokhale — CHANGES_REQUESTED** [review]
Some questions / minor comments.

File: presto-main/src/main/java/com/facebook/presto/connector/informationSchema/InformationSchemaMetadata.java (hunk 5 above)
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/connector/informationSchema/InformationSchemaMetadata.java:108]: Due to the SQL standard it is VARCHAR, and it takes YES/NO string values.

File: presto-hive/src/main/java/com/facebook/presto/hive/metastore/HivePrincipal.java (hunk 3 above)
Comment (arhimondr) [reply presto-hive/src/main/java/com/facebook/presto/hive/metastore/HivePrincipal.java:41]: PrestoPrincipal takes care of this.

File: presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java [hunk 6]
```diff
@@ -1791,6 +1818,16 @@ private static String typeParameterToString(SqlBaseParser.TypeParameterContext t
         throw new IllegalArgumentException("Unsupported typeParameter: " + typeParameter.getText());
     }
 
+    private static List<String> getIdentifiers(List<SqlBaseParser.IdentifierContext> identifiers)
+    {
+        return identifiers.stream().map(SqlBaseParser.IdentifierContext::getText).collect(toList());
```
Comment (maciejgrzybek) [inline presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java:1839]: `collect(toImmutableList())` instead of `toList()`.
We have a set of collectors in `com.facebook.presto.util.ImmutableCollectors`.
Copy in `visit...` will be no-op then anyway, if you were concerned of performance penalty.

File: presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java [hunk 7]
```diff
@@ -1791,6 +1818,16 @@ private static String typeParameterToString(SqlBaseParser.TypeParameterContext t
...
+    }
+
+    private static List<PrincipalSpecification> getPrincipalSpecifications(List<SqlBaseParser.PrincipalContext> principals)
+    {
+        return principals.stream().map(AstBuilder::getPrincipalSpecification).collect(toList());
```
Comment (maciejgrzybek) [inline presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java:1844]: ditto

File: presto-hive/src/main/java/com/facebook/presto/hive/security/SqlStandardAccessControl.java (hunk 1 above)
Comment (arhimondr) [reply presto-hive/src/main/java/com/facebook/presto/hive/security/SqlStandardAccessControl.java:467]: PrestoPrincipal does the lowercase

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/GrantRoles.java [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+    private GrantRoles(
+            Optional<NodeLocation> location,
+            Set<String> roles,
+            Set<PrincipalSpecification> grantees,
+            boolean withAdminOption,
+            Optional<GrantorSpecification> grantor,
+            Optional<String> catalog)
+    {
+        super(location);
+        this.roles = ImmutableSet.copyOf(requireNonNull(roles, "roles is null").stream().map(r -> r.toLowerCase(ENGLISH)).collect(toSet()));
```
Comment (maciejgrzybek) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/GrantRoles.java:68]: ditto (`collect(toImmutableSet())` instead of `ImmutableSet.copyOf(... toSet())`.

File: presto-hive/src/main/java/com/facebook/presto/hive/metastore/ThriftHiveMetastore.java (hunk 2 above)
Comment (arhimondr) [reply presto-hive/src/main/java/com/facebook/presto/hive/metastore/ThriftHiveMetastore.java]: I've taken care of this in `Prepare HiveMetastore interface to accept ROLE for GRANT/REVOKE` commit

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/RevokeRoles.java [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+    private RevokeRoles(
+            Optional<NodeLocation> location,
+            Set<String> roles,
+            Set<PrincipalSpecification> grantees,
+            boolean adminOptionFor,
+            Optional<GrantorSpecification> grantor,
+            Optional<String> catalog)
+    {
+        super(location);
+        this.roles = ImmutableSet.copyOf(requireNonNull(roles, "roles is null").stream().map(r -> r.toLowerCase(ENGLISH)).collect(toSet()));
```
Comment (maciejgrzybek) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/RevokeRoles.java:68]: ditto

File: presto-parser/src/test/java/com/facebook/presto/sql/parser/TestSqlParser.java [hunk 10]
```diff
@@ -1836,6 +1839,124 @@ public void testDropRole()
...
+        assertStatement("GRANT role1 TO ROLE role2 WITH ADMIN OPTION GRANTED BY ROLE admin IN catalog",
+                new GrantRoles(
+                        ImmutableSet.of("role1"),
+                        ImmutableSet.of(new PrincipalSpecification(PrincipalSpecification.Type.ROLE, "role2")),
+                        true,
+                        Optional.of(new GrantorSpecification(
+                                GrantorSpecification.Type.PRINCIPAL,
+                                Optional.of(new PrincipalSpecification(PrincipalSpecification.Type.ROLE, "admin")))),
+                        Optional.of("catalog")));
+    }
```
Comment (maciejgrzybek) [inline presto-parser/src/test/java/com/facebook/presto/sql/parser/TestSqlParser.java:1918]: Add test for `GRANT role to ROLE GRANTED BY` without `WITH ADMIN OPTION` part, given it's supported by parser.

File: presto-main/src/main/java/com/facebook/presto/execution/RevokeRolesTask.java [hunk 11]
```diff
@@ -0,0 +1,87 @@
+/*
```
Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/execution/RevokeRolesTask.java:1]: Remove that.

File: presto-main/src/main/java/com/facebook/presto/execution/GrantRolesTask.java [hunk 12]
```diff
@@ -0,0 +1,87 @@
+/*
```
Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/execution/GrantRolesTask.java:1]: Remove that.

File: presto-main/src/main/java/com/facebook/presto/execution/RevokeRolesTask.java [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+
+        Set<String> roles = statement.getRoles();
+        Set<PrestoPrincipal> grantees = statement.getGrantees().stream()
+                .map(MetadataUtil::createPrincipal)
+                .collect(toImmutableSet());
+        boolean adminOptionFor = statement.isAdminOptionFor();
+        Optional<PrestoPrincipal> grantor = statement.getGrantor().map(specification -> createPrincipal(session, specification));
+        String catalog = createCatalogName(session, statement, statement.getCatalog());
+
+        Set<String> availableRoles = metadata.listRoles(session, catalog);
```
Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/execution/RevokeRolesTask.java:61]: Don't you want to extract the check for existence of roles to some utility class?
E.g. `boolean doRolesExist(Collection<String> roles, Metadata metadata)`.
This code is exactly the same in both revoke and grant executions and is not trivial.

File: presto-spi/src/main/java/com/facebook/presto/spi/security/RoleGrant.java [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+public class RoleGrant
+{
+    private final PrestoPrincipal grantee;
+    private final String roleName;
+    private final boolean grantable;
+
+    @JsonCreator
+    public RoleGrant(@JsonProperty("grantee") PrestoPrincipal grantee, @JsonProperty("roleName") String roleName, @JsonProperty("grantable") boolean grantable)
+    {
+        this.grantee = requireNonNull(grantee, "grantee is null");
```
Comment (maciejgrzybek) [inline presto-spi/src/main/java/com/facebook/presto/spi/security/RoleGrant.java:33]: Is it case-sensitive?

File: presto-hive/src/main/java/com/facebook/presto/hive/ThriftHiveMetastoreClient.java [hunk 15]
```diff
@@ -239,4 +246,88 @@ public boolean revokePrivileges(PrivilegeBag privilegeBag)
...
+        request.setRequestType(GrantRevokeType.GRANT);
+        request.setRoleName(role);
+        request.setPrincipalName(granteeName);
+        request.setPrincipalType(granteeType);
+        request.setGrantor(grantorName);
+        request.setGrantorType(grantorType);
+        request.setGrantOption(grantOption);
+        GrantRevokeRoleResponse response = client.grant_revoke_role(request);
+        if (!response.isSetSuccess()) {
+            throw new MetaException("GrantRevokeResponse missing success field");
```
Comment (maciejgrzybek) [inline presto-hive/src/main/java/com/facebook/presto/hive/ThriftHiveMetastoreClient.java:275]: Is it guaranteed that role was not granted in `client.grant_revoke_roles` when `success` field is not set?
Don't we need some cleanup here?

File: presto-hive/src/main/java/com/facebook/presto/hive/ThriftHiveMetastoreClient.java [hunk 16]
```diff
@@ -239,4 +246,88 @@ public boolean revokePrivileges(PrivilegeBag privilegeBag)
...
+    @Override
+    public void revokeRole(String role, String granteeName, PrincipalType granteeType, boolean grantOption)
+            throws TException
+    {
+        List<RolePrincipalGrant> grants = listRoleGrants(granteeName, granteeType);
+        RolePrincipalGrant currentGrant = null;
+        for (RolePrincipalGrant grant : grants) {
+            if (grant.getRoleName().equals(role)) {
+                currentGrant = grant;
+                break;
```
Comment (maciejgrzybek) [inline presto-hive/src/main/java/com/facebook/presto/hive/ThriftHiveMetastoreClient.java:288]: Is it possible to have duplicates? If so, shouldn't we revoke all of them?

File: presto-hive/src/main/java/com/facebook/presto/hive/ThriftHiveMetastoreClient.java [hunk 17]
```diff
@@ -239,4 +246,88 @@ public boolean revokePrivileges(PrivilegeBag privilegeBag)
...
+    {
+        GrantRevokeRoleRequest request = new GrantRevokeRoleRequest();
+        request.setRequestType(GrantRevokeType.REVOKE);
+        request.setRoleName(role);
+        request.setPrincipalName(granteeName);
+        request.setPrincipalType(granteeType);
+        request.setGrantOption(grantOption);
+        GrantRevokeRoleResponse response = client.grant_revoke_role(request);
+        if (!response.isSetSuccess()) {
+            throw new MetaException("GrantRevokeResponse missing success field");
```
Comment (maciejgrzybek) [inline presto-hive/src/main/java/com/facebook/presto/hive/ThriftHiveMetastoreClient.java:314]: ditto

File: presto-hive/src/main/java/com/facebook/presto/hive/ThriftHiveMetastoreClient.java [hunk 18]
```diff
@@ -239,4 +246,88 @@ public boolean revokePrivileges(PrivilegeBag privilegeBag)
...
+        }
+    }
+
+    @Override
+    public List<RolePrincipalGrant> listRoleGrants(String principalName, PrincipalType principalType)
+            throws TException
+    {
+        GetRoleGrantsForPrincipalRequest request = new GetRoleGrantsForPrincipalRequest(principalName, principalType);
+        GetRoleGrantsForPrincipalResponse resp = client.get_role_grants_for_principal(request);
+        return resp.getPrincipalGrants();
```
Comment (maciejgrzybek) [inline presto-hive/src/main/java/com/facebook/presto/hive/ThriftHiveMetastoreClient.java]: `ImmutableList.copyOf`

File: presto-hive/src/main/java/com/facebook/presto/hive/metastore/MetastoreUtil.java [hunk 19]
```diff
@@ -344,6 +347,31 @@ public static Partition fromMetastoreApiPartition(org.apache.hadoop.hive.metasto
         return partitionBuilder.build();
     }
 
+    public static Set<RoleGrant> fromRolePrincipalGrants(Collection<RolePrincipalGrant> grants)
+    {
+        return ImmutableSet.copyOf(grants.stream().map(MetastoreUtil::fromRolePrincipalGrant).collect(toList()));
```
Comment (maciejgrzybek) [inline presto-hive/src/main/java/com/facebook/presto/hive/metastore/MetastoreUtil.java]: `toImmutableSet()` instead of `ImmutableSet.copyOf(... toList())`.

File: presto-hive/src/main/java/com/facebook/presto/hive/metastore/ThriftHiveMetastore.java [hunk 20]
```diff
@@ -286,6 +291,96 @@ public void dropRole(String role)
...
+    }
+
+    @Override
+    public Set<RoleGrant> listRoleGrants(PrestoPrincipal principal)
+    {
+        try {
+            return retry()
+                    .stopOn(MetaException.class)
+                    .stopOnIllegalExceptions()
+                    .run("revokeRole", stats.getListRoles().wrap(() -> {
```
Comment (maciejgrzybek) [inline presto-hive/src/main/java/com/facebook/presto/hive/metastore/ThriftHiveMetastore.java]: Not `revokeRole` here. It's `listRoleGrants` method.

File: presto-spi/src/main/java/com/facebook/presto/spi/security/RoleGrant.java [hunk 21]
```diff
@@ -0,0 +1,86 @@
...
+import static java.util.Objects.requireNonNull;
+
+public class RoleGrant
+{
+    private final PrestoPrincipal grantee;
+    private final String roleName;
+    private final boolean grantable;
+
+    @JsonCreator
+    public RoleGrant(@JsonProperty("grantee") PrestoPrincipal grantee, @JsonProperty("roleName") String roleName, @JsonProperty("grantable") boolean grantable)
```
Comment (maciejgrzybek) [inline presto-spi/src/main/java/com/facebook/presto/spi/security/RoleGrant.java:31]: Use enum instead of boolean for `grantable`.
It's non clear what's the intention when you use `new RoleGrant("...", "...", false)`
`new RoleGrant("...", "...", GRANTABLE)` or `new RoleGrant("...", "...", NON_GRANTABLE)` would be much more readable.

File: presto-hive/src/main/java/com/facebook/presto/hive/metastore/file/FileHiveMetastore.java [hunk 22]
```diff
@@ -578,6 +588,128 @@ public void dropRole(String role)
...
+        }
+    }
+
+    @Override
+    public Set<RoleGrant> listRoleGrants(PrestoPrincipal principal)
+    {
+        synchronized (rolesLock) {
+            return ImmutableSet.copyOf(listRoleGrantsSanitized().stream()
+                    .filter(grant -> grant.getGrantee().equals(principal))
+                    .collect(toSet()));
```
Comment (maciejgrzybek) [inline presto-hive/src/main/java/com/facebook/presto/hive/metastore/file/FileHiveMetastore.java]: `toImmutableSet`

File: presto-hive/src/main/java/com/facebook/presto/hive/metastore/file/FileHiveMetastore.java [hunk 23]
```diff
@@ -578,6 +588,128 @@ public void dropRole(String role)
...
+        }
+    }
+
+    private Set<RoleGrant> listRoleGrantsSanitized()
```
Comment (maciejgrzybek) [inline presto-hive/src/main/java/com/facebook/presto/hive/metastore/file/FileHiveMetastore.java]: This method is misleading. It suggests is pure getter with some postprocessing but in fact it may change the content of the file (when file is not sanitized). I'd prefer having a separate method for it.

File: presto-hive/src/main/java/com/facebook/presto/hive/security/SqlStandardAccessControl.java [hunk 24]
```diff
@@ -331,15 +350,74 @@ private boolean checkTablePermission(ConnectorTransactionHandle transaction, Ide
...
+        return MetastoreUtil.listApplicableRoles(principal, metastore::listRoleGrants)
+                .stream()
+                .map(RoleGrant::getRoleName)
+                .collect(toSet());
+    }
+
+    private static Set<HivePrivilegeInfo> listApplicableTablePrivileges(SemiTransactionalHiveMetastore metastore, String databaseName, String tableName, PrestoPrincipal principal)
+    {
+        Set<String> applicableRoles = listApplicableRoles(metastore, principal);
+        List<PrestoPrincipal> principals = new ArrayList<>();
```
Comment (maciejgrzybek) [inline presto-hive/src/main/java/com/facebook/presto/hive/security/SqlStandardAccessControl.java:465]: Why not `ImmutableList.Builder`?

File: presto-hive/src/main/java/com/facebook/presto/hive/security/SqlStandardAccessControl.java [hunk 25]
```diff
@@ -313,6 +313,47 @@ public void checkCanDropRole(ConnectorTransactionHandle transactionHandle, Ident
...
+
+    private boolean hasAdminOption(ConnectorTransactionHandle transaction, Identity identity, Set<String> roles)
+    {
+        SemiTransactionalHiveMetastore metastore = metastoreProvider.apply(((HiveTransactionHandle) transaction));
+        Set<RoleGrant> grants = MetastoreUtil.listApplicableRoles(new PrestoPrincipal(USER, identity.getUser()), metastore::listRoleGrants);
+        Set<String> rolesWithGrantOption = grants.stream()
+                .filter(RoleGrant::isGrantable)
+                .map(RoleGrant::getRoleName)
+                .collect(toSet());
+        return rolesWithGrantOption.containsAll(roles);
```
Comment (maciejgrzybek) [inline presto-hive/src/main/java/com/facebook/presto/hive/security/SqlStandardAccessControl.java:364]: What if `roles` is a superset of `rolesWithGrantOption`? Should the user be recognized as "hasAdminOption", then?

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java [hunk 26]
```diff
@@ -1284,7 +1285,18 @@ public void revokeTablePrivileges(ConnectorSession session, SchemaTableName sche
                 .map(privilege -> new HivePrivilegeInfo(toHivePrivilege(privilege), grantOption))
                 .collect(toSet());
 
-        metastore.revokeTablePrivileges(schemaName, tableName, grantee, hivePrivilegeInfos);
+        metastore.revokeTablePrivileges(schemaName, tableName, getPrestoPrincipal(grantee), hivePrivilegeInfos);
+    }
+
+    private PrestoPrincipal getPrestoPrincipal(String grantee)
+    {
+        // TODO this hack will be removed after grant for roles in introduced
```
Comment (maciejgrzybek) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: typo: `is`

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java [hunk 27]
```diff
@@ -1847,6 +1849,211 @@ public void testDropNonExistentRole()
...
+                .build();
+        MaterializedResult actual = queryRunner.execute(session, "SELECT * FROM hive.information_schema.applicable_roles");
+        MaterializedResult expected = MaterializedResult.resultBuilder(session, createUnboundedVarcharType(), createUnboundedVarcharType(), createUnboundedVarcharType(), createUnboundedVarcharType())
+                .row("admin_user", "USER", "admin", "NO")
+                .build();
+        assertContains(actual, expected);
+    }
+
+    @Test
+    public void testGrantRevokeRole()
```
Comment (maciejgrzybek) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java]: Break this test into smaller pieces. It's hard to follow it. Especially given comparison of `actual` and `expected` is pretty verbose (8-10 lines each block).

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java [hunk 28]
```diff
@@ -1847,6 +1849,211 @@ public void testDropNonExistentRole()
...
+            throws Exception
+    {
+        assertUpdate("CREATE ROLE grant_revoke_role_1");
```
Comment (maciejgrzybek) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java]: Once broken into smaller tests, these could be in a common setup method.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java [hunk 29]
```diff
@@ -1847,6 +1849,211 @@ public void testDropNonExistentRole()
...
+
+        MaterializedResult actual = queryRunner.execute(user1, "SELECT * FROM hive.information_schema.applicable_roles");
+        MaterializedResult expected = MaterializedResult.resultBuilder(user1, createUnboundedVarcharType(), createUnboundedVarcharType(), createUnboundedVarcharType(), createUnboundedVarcharType())
+                .row("grant_revoke_user_1", "USER", "admin", "NO")
+                .row("grant_revoke_user_1", "USER", "public", "NO")
+                .row("grant_revoke_user_1", "USER", "grant_revoke_role_1", "NO")
+                .row("grant_revoke_role_1", "ROLE", "grant_revoke_role_2", "NO")
+                .row("grant_revoke_role_1", "ROLE", "grant_revoke_role_3", "NO")
+                .build();
+        assertEqualsIgnoreOrder(actual, expected);
```
Comment (maciejgrzybek) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java]: This would be the end of the first test - "grantedRolesAreRetrieved".

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java [hunk 30]
```diff
@@ -1847,6 +1849,211 @@ public void testDropNonExistentRole()
...
+        assertUpdate("REVOKE grant_revoke_role_3 FROM ROLE grant_revoke_role_1");
+
+        actual = queryRunner.execute(user1, "SELECT * FROM hive.information_schema.applicable_roles");
+        expected = MaterializedResult.resultBuilder(user1, createUnboundedVarcharType(), createUnboundedVarcharType(), createUnboundedVarcharType(), createUnboundedVarcharType())
+                .row("grant_revoke_user_1", "USER", "admin", "NO")
+                .row("grant_revoke_user_1", "USER", "public", "NO")
+                .row("grant_revoke_user_1", "USER", "grant_revoke_role_1", "NO")
+                .row("grant_revoke_role_1", "ROLE", "grant_revoke_role_2", "NO")
+                .build();
+        assertEqualsIgnoreOrder(actual, expected);
```
Comment (maciejgrzybek) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java]: This would be second - "afterRevokingRoleIsNoLongerListed".

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java [hunk 31]
```diff
@@ -1847,6 +1849,211 @@ public void testDropNonExistentRole()
...
+        expected = MaterializedResult.resultBuilder(user1, createUnboundedVarcharType(), createUnboundedVarcharType(), createUnboundedVarcharType(), createUnboundedVarcharType())
+                .row("grant_revoke_user_1", "USER", "admin", "NO")
+                .row("grant_revoke_user_1", "USER", "public", "NO")
+                .row("grant_revoke_user_1", "USER", "grant_revoke_role_1", "NO")
+                .row("grant_revoke_role_1", "ROLE", "grant_revoke_role_2", "NO")
+                .row("grant_revoke_role_1", "ROLE", "grant_revoke_role_3", "NO")
+                .build();
+        assertEqualsIgnoreOrder(actual, expected);
+
+        assertUpdate("DROP ROLE grant_revoke_role_3");
```
Comment (maciejgrzybek) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java]: etc...

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java [hunk 32]
```diff
@@ -1847,6 +1849,211 @@ public void testDropNonExistentRole()
...
+                .row("grant_revoke_user_1", "USER", "admin", "NO")
+                .row("grant_revoke_user_1", "USER", "public", "NO")
+                .build();
+        assertEqualsIgnoreOrder(actual, expected);
+
+        assertUpdate("DROP ROLE grant_revoke_role_2");
+    }
+
+    @Test
+    public void testGrantRoleWithGrantOption()
```
Comment (maciejgrzybek) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java]: Ditto. Break this into smaller tests to better express your intention, what you exactly test and to emphasize isolation (not depending on side effects of previous tests).

File: presto-product-tests/src/main/java/com/facebook/presto/tests/hive/TestRoles.java [hunk 33]
```diff
@@ -137,8 +138,198 @@ public void testAccessControl()
...
+    }
+
+    @Test(groups = {HIVE_CONNECTOR, ROLES, AUTHORIZATION, PROFILE_SPECIFIC_TESTS})
+    public void testGrantRevokeRole()
```
Comment (maciejgrzybek) [inline presto-product-tests/src/main/java/com/facebook/presto/tests/hive/TestRoles.java]: Ditto break test into shorter tests.

File: presto-product-tests/src/main/java/com/facebook/presto/tests/hive/TestRoles.java [hunk 34]
```diff
@@ -137,8 +138,198 @@ public void testAccessControl()
...
+                        row("role1", "ROLE", "role2", "NO"));
+
+        onPresto().executeQuery("DROP ROLE role1");
+        QueryAssert.assertThat(onPrestoAlice().executeQuery("SELECT * FROM hive.information_schema.applicable_roles"))
+                .containsOnly(
+                        row("alice", "USER", "public", "NO"));
+    }
+
+    @Test(groups = {HIVE_CONNECTOR, ROLES, AUTHORIZATION, PROFILE_SPECIFIC_TESTS})
+    public void testGrantRoleWithGrantOption()
```
Comment (maciejgrzybek) [inline presto-product-tests/src/main/java/com/facebook/presto/tests/hive/TestRoles.java]: ditto

**Review by maciejgrzybek — CHANGES_REQUESTED** [review]
3b89ba62ef07566d9c1febd08d924602021ad956 should be squashed with 2c6dc4d1d50a3427099bf095ca925be1c1d11e00.

5f1d6ef0d6ba165741161bbe97682d6fcfceff62 should be squashed with 7c30f5d0c619e5ba98783ed3ad1755b3bd1fabb2.

File: presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java (hunk 6 above)
Comment (arhimondr) [reply presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java:1839]: `toImmutableList()` is not available in `presto-parser` module. `toList()` collector is used all across the code of this class.

File: presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java (hunk 7 above)
Comment (arhimondr) [reply presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java:1844]: ditto

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/RevokeRoles.java (hunk 9 above)
Comment (arhimondr) [reply presto-parser/src/main/java/com/facebook/presto/sql/tree/RevokeRoles.java:68]: ditto

File: presto-main/src/main/java/com/facebook/presto/execution/RevokeRolesTask.java (hunk 13 above)
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/execution/RevokeRolesTask.java:61]: They you will either extract 3 lines, by cost of creation new class, or, if you want to extract the entire check code you need to introduce additional abstractions, like `RoleManipulationTask` abstract task that returns all the available roles.

File: presto-spi/src/main/java/com/facebook/presto/spi/security/RoleGrant.java (hunk 21 above)
Comment (arhimondr) [reply presto-spi/src/main/java/com/facebook/presto/spi/security/RoleGrant.java:31]: Yeah, in some places boolean is not that readable. It can be overcome with introducing builder. But actually we create this class only in several places, and only 2 of them are a bit consusing. All the other places are pretty easy to interpret.

File: presto-spi/src/main/java/com/facebook/presto/spi/security/RoleGrant.java (hunk 14 above)
Comment (arhimondr) [reply presto-spi/src/main/java/com/facebook/presto/spi/security/RoleGrant.java:33]: It is case insensitive. `toLowerCase` is being made within the `PrestoPrincipal` constructor.

File: presto-hive/src/main/java/com/facebook/presto/hive/ThriftHiveMetastoreClient.java (hunk 15 above)
Comment (arhimondr) [reply presto-hive/src/main/java/com/facebook/presto/hive/ThriftHiveMetastoreClient.java:275]: I copied it literally from hive source. That's how they check that the grant was successful.

File: presto-hive/src/main/java/com/facebook/presto/hive/ThriftHiveMetastoreClient.java (hunk 16 above)
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/ThriftHiveMetastoreClient.java]: With hive you never no. You are right. It is better to check the entire list.

Comment (arhimondr) [reply presto-hive/src/main/java/com/facebook/presto/hive/ThriftHiveMetastoreClient.java:288]: You can't revoke some particular grant object. You can only revoke by grantee-role pair. So, it doesn't make sense to call the same method with the same parameters twice.

File: presto-hive/src/main/java/com/facebook/presto/hive/ThriftHiveMetastoreClient.java (hunk 17 above)
Comment (arhimondr) [reply presto-hive/src/main/java/com/facebook/presto/hive/ThriftHiveMetastoreClient.java:314]: ditto

File: presto-hive/src/main/java/com/facebook/presto/hive/metastore/MetastoreUtil.java (hunk 19 above)
Comment (arhimondr) [reply presto-hive/src/main/java/com/facebook/presto/hive/metastore/MetastoreUtil.java]: It is not available in this module

File: presto-hive/src/main/java/com/facebook/presto/hive/metastore/file/FileHiveMetastore.java (hunk 22 above)
Comment (arhimondr) [reply presto-hive/src/main/java/com/facebook/presto/hive/metastore/file/FileHiveMetastore.java]: it is not available in this module

File: presto-hive/src/main/java/com/facebook/presto/hive/security/SqlStandardAccessControl.java (hunk 25 above)
Comment (arhimondr) [reply presto-hive/src/main/java/com/facebook/presto/hive/security/SqlStandardAccessControl.java:364]: we just check that set of roles that we can grant contains all the roles we are trying to grant

File: presto-hive/src/main/java/com/facebook/presto/hive/security/SqlStandardAccessControl.java (hunk 24 above)
Comment (arhimondr) [reply presto-hive/src/main/java/com/facebook/presto/hive/security/SqlStandardAccessControl.java:465]: This is just a local variable. I'm not sure if it worth extra code to change it to builder.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java (hunk 28 above)
Comment (arhimondr) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java]: I cannot move it to the setup method. This test class may be executed concurrently.

File: presto-hive/src/main/java/com/facebook/presto/hive/ThriftHiveMetastoreClient.java (hunk 15 above)
Comment (maciejgrzybek) [reply presto-hive/src/main/java/com/facebook/presto/hive/ThriftHiveMetastoreClient.java:275]: The question is if we need some extra cleanup if `client.grant_revoke_role` failed. Presumably not?

**Review by maciejgrzybek — APPROVED** [review]
Fire in the hole!

**Review by amrutagokhale — APPROVED** [review]
